![HR Analytics](../Images/Data_Preperation.png)

---

**_The key to success in any organization lies in its ability to attract and retain top talent. As part of the Unified Mentor program, I undertook the IBM HR Analytics Employee Attrition & Performance project to analyze critical factors influencing employee turnover._**

**_This analysis is designed to support HR Analysts in identifying the underlying reasons why employees choose to stay or leave an organization. By leveraging data-driven insights, the project highlights patterns related to job satisfaction, work-life balance, compensation, and other key variables impacting attrition._**

**_Understanding these factors empowers HR teams to implement proactive strategies aimed at improving employee retention, enhancing workplace satisfaction, and ultimately reducing the costly cycle of hiring and training new staff. The predictive models and visual insights developed in this project serve as valuable tools to help organizations retain their top performers and foster a stable, productive workforce._**

---
# <span style="background-color:#393be5; color:white; padding:10px;border-radius:15px; text-align:center;">OUTLINE</span>

1. **Import Libraries**  
   Load the necessary Python tools to work with data and perform analysis.

2. **Load the Data**  
   Open the dataset and get it ready for analysis.

3. **Explore and Clean the Data (Data Wrangling)**  
   - Understanding Dataset Dimensions  
   - Exploring Dataset Structure  
   - Attribute Summary  
   - Feature Classification  
   - Decoding Encoded Features  
   - Look for any missing or empty data.  
   - Summarize the numeric data to spot trends (like average salary, age range).  
   - Remove columns that aren’t useful for analysis.  
   - Review category-based columns to understand common values (e.g., most common job role).  
   - Check how many unique values exist in each category column.

4. **Save the Cleaned Data**  
   Store the cleaned dataset in a new file for future use.

---

# <span style="background-color:#258683; color:white; padding:10px;border-radius:15px; text-align:center;">IMPORTING VARIOUS MODULES</span>


In [1]:
# Library for Data Manipulation
import numpy as np
import pandas as pd
import io
# Library for Statistical Modelling
from sklearn.preprocessing import LabelEncoder
# Library for Ignore the warnings
import warnings
warnings.filterwarnings('always')
warnings.filterwarnings('ignore')

---
# <span style="background-color:#258683; color:white; padding:10px;border-radius:15px; text-align:center;">PROVIDED DATASET</span>

In [2]:
# Load the dataset
employee_data = '../WA_Fn-UseC_-HR-Employee-Attrition.csv'
df = pd.read_csv(employee_data)

# Display the first 5 rows
df.head()

Age Attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
3   33        No  Travel_Frequently       1392  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   
3                 3          4  Life Sciences              1               5   
4                 2          1        Medical              1               7   

   ...  RelationshipSatisfaction StandardHours  StockOptionLevel  \
0  ...                         1            80                 0   
1  ...                         4            80                 1   
2  ...                         2            80                 0   
3  ...                         3            80                 0   
4  ...                         4            80                 1   

   TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
0                  8                      0               1               6   
1                 10                      3               3              10   
2                  7                      3               3               0   
3                  8                      3               3               8   
4                  6                      3               3               2   

  YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
0                  4                        0                     5  
1                  7                        1                     7  
2                  0                        0                     0  
3                  7                        3                     0  
4                  2                        2                     2  

[5 rows x 35 columns]

In [3]:
df.tail()

Age Attrition     BusinessTravel  DailyRate              Department  \
1465   36        No  Travel_Frequently        884  Research & Development   
1466   39        No      Travel_Rarely        613  Research & Development   
1467   27        No      Travel_Rarely        155  Research & Development   
1468   49        No  Travel_Frequently       1023                   Sales   
1469   34        No      Travel_Rarely        628  Research & Development   

      DistanceFromHome  Education EducationField  EmployeeCount  \
1465                23          2        Medical              1   
1466                 6          1        Medical              1   
1467                 4          3  Life Sciences              1   
1468                 2          3        Medical              1   
1469                 8          3        Medical              1   

      EmployeeNumber  ...  RelationshipSatisfaction StandardHours  \
1465            2061  ...                         3            80   
1466            2062  ...                         1            80   
1467            2064  ...                         2            80   
1468            2065  ...                         4            80   
1469            2068  ...                         1            80   

      StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
1465                 1                 17                      3   
1466                 1                  9                      5   
1467                 1                  6                      0   
1468                 0                 17                      3   
1469                 0                  6                      3   

     WorkLifeBalance  YearsAtCompany YearsInCurrentRole  \
1465               3               5                  2   
1466               3               7                  7   
1467               3               6                  2   
1468               2               9                  6   
1469               4               4                  3   

      YearsSinceLastPromotion  YearsWithCurrManager  
1465                        0                     3  
1466                        1                     7  
1467                        0                     3  
1468                        0                     8  
1469                        1                     2  

[5 rows x 35 columns]

---
# <span style="background-color:#258683; color:white; padding:10px;border-radius:15px; text-align:center;">Understanding Dataset Dimensions</span>

In [4]:
# Checking the shape of the dataset
rows, columns = df.shape

rows, columns

(1470, 35)

---
# <span style="background-color:#258683; color:white; padding:10px;border-radius:15px; text-align:center;">Exploring Dataset Structure</span>

In [5]:
# Listing all columns in the dataset
df.columns.tolist()


['Age',
 'Attrition',
 'BusinessTravel',
 'DailyRate',
 'Department',
 'DistanceFromHome',
 'Education',
 'EducationField',
 'EmployeeCount',
 'EmployeeNumber',
 'EnvironmentSatisfaction',
 'Gender',
 'HourlyRate',
 'JobInvolvement',
 'JobLevel',
 'JobRole',
 'JobSatisfaction',
 'MaritalStatus',
 'MonthlyIncome',
 'MonthlyRate',
 'NumCompaniesWorked',
 'Over18',
 'OverTime',
 'PercentSalaryHike',
 'PerformanceRating',
 'RelationshipSatisfaction',
 'StandardHours',
 'StockOptionLevel',
 'TotalWorkingYears',
 'TrainingTimesLastYear',
 'WorkLifeBalance',
 'YearsAtCompany',
 'YearsInCurrentRole',
 'YearsSinceLastPromotion',
 'YearsWithCurrManager']

---
# <span style="background-color:#258683; color:white; padding:10px;border-radius:15px; text-align:center;">Attribute Summary</span>

**We’ll generate basic info:**
- Data types
- Non-null counts
- Memory usage

In [6]:
# Generating basic information about the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   object
 2   BusinessTravel            1470 non-null   object
 3   DailyRate                 1470 non-null   int64 
 4   Department                1470 non-null   object
 5   DistanceFromHome          1470 non-null   int64 
 6   Education                 1470 non-null   int64 
 7   EducationField            1470 non-null   object
 8   EmployeeCount             1470 non-null   int64 
 9   EmployeeNumber            1470 non-null   int64 
 10  EnvironmentSatisfaction   1470 non-null   int64 
 11  Gender                    1470 non-null   object
 12  HourlyRate                1470 non-null   int64 
 13  JobInvolvement            1470 non-null   int64 
 14  JobLevel                

***Attribute Summary:***
- Total Entries: 1470
- Total Features: 35

**Data Types:**
- 26 Numerical (int64)
- 9 Categorical (object)

**No missing values detected so far (all columns show 1470 non-null).**


---
# <span style="background-color:#258683; color:white; padding:10px;border-radius:15px; text-align:center;">Feature Classification</span>
- List all Numerical Features.

- List all Categorical Features.

In [7]:
# Identifying Numerical and Categorical Features
numerical_features = df.select_dtypes(include=['int64', 'float64']).columns.tolist()
categorical_features = df.select_dtypes(include=['object']).columns.tolist()

numerical_features, categorical_features

(['Age',
  'DailyRate',
  'DistanceFromHome',
  'Education',
  'EmployeeCount',
  'EmployeeNumber',
  'EnvironmentSatisfaction',
  'HourlyRate',
  'JobInvolvement',
  'JobLevel',
  'JobSatisfaction',
  'MonthlyIncome',
  'MonthlyRate',
  'NumCompaniesWorked',
  'PercentSalaryHike',
  'PerformanceRating',
  'RelationshipSatisfaction',
  'StandardHours',
  'StockOptionLevel',
  'TotalWorkingYears',
  'TrainingTimesLastYear',
  'WorkLifeBalance',
  'YearsAtCompany',
  'YearsInCurrentRole',
  'YearsSinceLastPromotion',
  'YearsWithCurrManager'],
 ['Attrition',
  'BusinessTravel',
  'Department',
  'EducationField',
  'Gender',
  'JobRole',
  'MaritalStatus',
  'Over18',
  'OverTime'])

***Numerical Features (26):***

- **Here are the key numerical columns:** ['Age', 'DailyRate', 'DistanceFromHome', 'Education', 'EmployeeCount', 'EmployeeNumber', 'EnvironmentSatisfaction', 'HourlyRate', 'JobInvolvement', 'JobLevel', 'JobSatisfaction', 'MonthlyIncome', 'MonthlyRate', 'NumCompaniesWorked', 'PercentSalaryHike', 'PerformanceRating', 'RelationshipSatisfaction', 'StandardHours', 'StockOptionLevel', 'TotalWorkingYears', 'TrainingTimesLastYear', 'WorkLifeBalance', 'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion', 'YearsWithCurrManager']

***Categorical Features (9):***

- **Here are the key categorical columns:** ['Attrition', 'BusinessTravel', 'Department', 'EducationField', 'Gender', 'JobRole', 'MaritalStatus', 'Over18', 'OverTime']

---
# <span style="background-color:#258683; color:white; padding:10px;border-radius:15px; text-align:center;">Decoding Encoded Features</span>

**Mappings to Apply:**
- Education {1: 'Below College', 2: 'College', 3: 'Bachelor', 4: 'Master', 5: 'Doctor'}
- EnvironmentSatisfaction, JobInvolvement, JobSatisfaction, RelationshipSatisfaction {1: 'Low', 2: 'Medium', 3: 'High', 4: 'Very High'}
- PerformanceRating {1: 'Low', 2: 'Good', 3: 'Excellent', 4: 'Outstanding'}
- WorkLifeBalance {1: 'Bad', 2: 'Good', 3: 'Better', 4: 'Best'}

In [8]:
# Defining the mappings
education_map = {1: 'Below College', 2: 'College', 3: 'Bachelor', 4: 'Master', 5: 'Doctor'}
satisfaction_map = {1: 'Low', 2: 'Medium', 3: 'High', 4: 'Very High'}
performance_map = {1: 'Low', 2: 'Good', 3: 'Excellent', 4: 'Outstanding'}
worklife_map = {1: 'Bad', 2: 'Good', 3: 'Better', 4: 'Best'}

# Applying mappings
df['Education'] = df['Education'].map(education_map)
df['EnvironmentSatisfaction'] = df['EnvironmentSatisfaction'].map(satisfaction_map)
df['JobInvolvement'] = df['JobInvolvement'].map(satisfaction_map)
df['JobSatisfaction'] = df['JobSatisfaction'].map(satisfaction_map)
df['RelationshipSatisfaction'] = df['RelationshipSatisfaction'].map(satisfaction_map)
df['PerformanceRating'] = df['PerformanceRating'].map(performance_map)
df['WorkLifeBalance'] = df['WorkLifeBalance'].map(worklife_map)

# Checking if mapping is successful
df[['Education', 'EnvironmentSatisfaction', 'JobInvolvement', 'JobSatisfaction', 
    'RelationshipSatisfaction', 'PerformanceRating', 'WorkLifeBalance']].head()


Education EnvironmentSatisfaction JobInvolvement JobSatisfaction  \
0        College                  Medium           High       Very High   
1  Below College                    High         Medium          Medium   
2        College               Very High         Medium            High   
3         Master               Very High           High            High   
4  Below College                     Low           High          Medium   

  RelationshipSatisfaction PerformanceRating WorkLifeBalance  
0                      Low         Excellent             Bad  
1                Very High       Outstanding          Better  
2                   Medium         Excellent          Better  
3                     High         Excellent          Better  
4                Very High         Excellent          Better

---
# <span style="background-color:#258683; color:white; padding:10px;border-radius:15px; text-align:center;">Validating Data Completeness</span>

- Next, let's check for any missing values across the dataset.

In [9]:
# Checking for missing values
missing_values = df.isnull().sum()

# Filtering columns with missing values
missing_values = missing_values[missing_values > 0]

missing_values

Series([], dtype: int64)

- All columns are complete—no null values in the dataset.

---
# <span style="background-color:#258683; color:white; padding:10px;border-radius:15px; text-align:center;">Statistical Summary of Numerical Data</span>

**We’ll run describe() to get:**

- Count
- Mean
- Std deviation
- Min & Max
- Percentiles

In [10]:
# Descriptive statistics for numerical features
numerical_df = df.select_dtypes(include=['int64', 'float64'])
numerical_df.describe()


Age    DailyRate  DistanceFromHome  EmployeeCount  \
count  1470.000000  1470.000000       1470.000000         1470.0   
mean     36.923810   802.485714          9.192517            1.0   
std       9.135373   403.509100          8.106864            0.0   
min      18.000000   102.000000          1.000000            1.0   
25%      30.000000   465.000000          2.000000            1.0   
50%      36.000000   802.000000          7.000000            1.0   
75%      43.000000  1157.000000         14.000000            1.0   
max      60.000000  1499.000000         29.000000            1.0   

       EmployeeNumber   HourlyRate     JobLevel  MonthlyIncome   MonthlyRate  \
count     1470.000000  1470.000000  1470.000000    1470.000000   1470.000000   
mean      1024.865306    65.891156     2.063946    6502.931293  14313.103401   
std        602.024335    20.329428     1.106940    4707.956783   7117.786044   
min          1.000000    30.000000     1.000000    1009.000000   2094.000000   
25%        491.250000    48.000000     1.000000    2911.000000   8047.000000   
50%       1020.500000    66.000000     2.000000    4919.000000  14235.500000   
75%       1555.750000    83.750000     3.000000    8379.000000  20461.500000   
max       2068.000000   100.000000     5.000000   19999.000000  26999.000000   

       NumCompaniesWorked  PercentSalaryHike  StandardHours  StockOptionLevel  \
count         1470.000000        1470.000000         1470.0       1470.000000   
mean             2.693197          15.209524           80.0          0.793878   
std              2.498009           3.659938            0.0          0.852077   
min              0.000000          11.000000           80.0          0.000000   
25%              1.000000          12.000000           80.0          0.000000   
50%              2.000000          14.000000           80.0          1.000000   
75%              4.000000          18.000000           80.0          1.000000   
max              9.000000          25.000000           80.0          3.000000   

       TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
count        1470.000000            1470.000000     1470.000000   
mean           11.279592               2.799320        7.008163   
std             7.780782               1.289271        6.126525   
min             0.000000               0.000000        0.000000   
25%             6.000000               2.000000        3.000000   
50%            10.000000               3.000000        5.000000   
75%            15.000000               3.000000        9.000000   
max            40.000000               6.000000       40.000000   

       YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
count         1470.000000              1470.000000           1470.000000  
mean             4.229252                 2.187755              4.123129  
std              3.623137                 3.222430              3.568136  
min              0.000000                 0.000000              0.000000  
25%              2.000000                 0.000000              2.000000  
50%              3.000000                 1.000000              3.000000  
75%              7.000000                 3.000000              7.000000  
max             18.000000                15.000000             17.000000

***Here’s the Statistical Summary of Numerical Features directly displayed:***

**You can observe key stats like mean, std, min, max, and quartiles for each numerical column.**

 **Notice columns like:**
 - StandardHours and EmployeeCount are constants.
 - EmployeeNumber is just an ID.

---
# <span style="background-color:#258683; color:white; padding:10px;border-radius:15px; text-align:center;">Streamlining the Dataset</span>

**Remove columns that don’t add value for analysis or modeling.**

**Columns to Drop:**
- Column---------------------------------------------------Reason
- EmployeeCount------------------------Constant value (always 1)
- Over18-------------------------Constant value ('Y' for all entries)
- StandardHours------------------------Constant value (always 80)
- EmployeeNumber----Just a unique ID, no predictive importance

**Dropping these will clean up the dataset and avoid noise in further steps like EDA or Modeling.**

In [11]:
# Dropping unnecessary columns
df.drop(['EmployeeCount', 'Over18', 'StandardHours', 'EmployeeNumber'], axis=1, inplace=True)

# Verifying the shape after dropping
df.shape

(1470, 31)

**We have successfully dropped unnecessary columns.**

- The dataset now contains 31 columns (down from 35).
- All remaining features are relevant for analysis.

---
# <span style="background-color:#258683; color:white; padding:10px;border-radius:15px; text-align:center;">Summary of Categorical Data</span>

**Run describe() specifically on the categorical features to understand:**
- How many unique categories each has.
- The most frequent category.
- How often that category appears.

In [12]:
# Ensuring proper selection of categorical features after mapping
categorical_columns = df.columns[df.dtypes == 'object']

# Descriptive summary for actual categorical columns
categorical_summary = df[categorical_columns].describe()

categorical_summary


Attrition BusinessTravel              Department Education  \
count       1470           1470                    1470      1470   
unique         2              3                       3         5   
top           No  Travel_Rarely  Research & Development  Bachelor   
freq        1233           1043                     961       572   

       EducationField EnvironmentSatisfaction Gender JobInvolvement  \
count            1470                    1470   1470           1470   
unique              6                       4      2              4   
top     Life Sciences                    High   Male           High   
freq              606                     453    882            868   

                JobRole JobSatisfaction MaritalStatus OverTime  \
count              1470            1470          1470     1470   
unique                9               4             3        2   
top     Sales Executive       Very High       Married       No   
freq                326             459           673     1054   

       PerformanceRating RelationshipSatisfaction WorkLifeBalance  
count               1470                     1470            1470  
unique                 2                        4               4  
top            Excellent                     High          Better  
freq                1244                      459             893

---
# <span style="background-color:#258683; color:white; padding:10px;border-radius:15px; text-align:center;">Exploring Categorical Diversity</span>

**Check unique values in each categorical feature to fully understand category distributions.**

In [13]:
# Checking unique values for each categorical attribute
unique_values = {col: df[col].unique() for col in categorical_columns}

unique_values

{'Attrition': array(['Yes', 'No'], dtype=object),
 'BusinessTravel': array(['Travel_Rarely', 'Travel_Frequently', 'Non-Travel'], dtype=object),
 'Department': array(['Sales', 'Research & Development', 'Human Resources'], dtype=object),
 'Education': array(['College', 'Below College', 'Master', 'Bachelor', 'Doctor'],
       dtype=object),
 'EducationField': array(['Life Sciences', 'Other', 'Medical', 'Marketing',
        'Technical Degree', 'Human Resources'], dtype=object),
 'EnvironmentSatisfaction': array(['Medium', 'High', 'Very High', 'Low'], dtype=object),
 'Gender': array(['Female', 'Male'], dtype=object),
 'JobInvolvement': array(['High', 'Medium', 'Very High', 'Low'], dtype=object),
 'JobRole': array(['Sales Executive', 'Research Scientist', 'Laboratory Technician',
        'Manufacturing Director', 'Healthcare Representative', 'Manager',
        'Sales Representative', 'Research Director', 'Human Resources'],
       dtype=object),
 'JobSatisfaction': array(['Very High', 'Mediu

**Unique Values in Categorical Features:**
- Attrition: ['Yes', 'No']
- BusinessTravel: ['Travel_Rarely', 'Travel_Frequently', 'Non-Travel']
- Department: ['Sales', 'Research & Development', 'Human Resources']
- Education: ['College', 'Below College', 'Master', 'Bachelor', 'Doctor']
- EducationField: ['Life Sciences', 'Other', 'Medical', 'Marketing', 'Technical Degree', 'Human Resources']
- EnvironmentSatisfaction: ['Medium', 'High', 'Very High', 'Low']
- Gender: ['Female', 'Male']
- JobInvolvement: ['High', 'Medium', 'Very High', 'Low']
- JobRole: ['Sales Executive', 'Research Scientist', 'Laboratory Technician', 'Manufacturing Director', 'Healthcare Representative', 'Manager', 'Sales Representative', 'Research Director', 'Human Resources']
- JobSatisfaction: ['Very High', 'Medium', 'High', 'Low']
- MaritalStatus: ['Single', 'Married', 'Divorced']
- OverTime: ['Yes', 'No']
- PerformanceRating: ['Excellent', 'Outstanding']
- RelationshipSatisfaction: ['Low', 'Very High', 'Medium', 'High']
- WorkLifeBalance: ['Bad', 'Better', 'Good', 'Best']

---
# <span style="background-color:#258683; color:white; padding:10px;border-radius:15px; text-align:center;">Saving the Refined Dataset</span>

**Save this cleaned and prepared dataset into a CSV file for future steps like EDA, Statistical Analysis, and Modeling.**

In [14]:
cleaned_data_path = '../CSV/employee_attrition_cleaned.csv'
df.to_csv(cleaned_data_path, index=False)

cleaned_data_path

'../CSV/employee_attrition_cleaned.csv'

---